packages/vats/src/transfer.js

```diff
@@ -0,0 +1,186 @@
+            Fail`Invalid upcall argument type ${obj.type}; expected VTRANSFER_IBC_EVENT`;
+
+          // First, call our target contract listener.
+          // A VTransfer active interceptor can return a write acknowledgement
+          /** @type {import('@agoric/vow').Vow<any>} */
+          let retP = watch(E(tap).upcall(obj));
+
+          // See if the upcall result needs special handling.
+          if (obj.event === 'writeAcknowledgement') {
+            const ackMethod = {
+              type: 'IBC_METHOD',
+              method: 'receiveExecuted',
+              packet: obj.packet,
+            };
+            if (isActiveTap) {
+              // FIXME: use the Vow `watch` operator.
+              retP = watch(retP, this.facets.encodeAckWatcher, { ackMethod });
+            } else {
+              // This is a passive tap, so forward the ack without intervention.
+              ackMethod.ack = obj.acknowledgement;
+              retP = watch(E(system).downcall(ackMethod));
+            }
+            retP = watch(retP, this.facets.sendErrorWatcher, { ackMethod });
+          }
+
+          // Log errors in the upcall handling.
+          retP = watch(retP, this.facets.logErrorWatcher, { obj });
+
+          if (isActiveTap) {
+            // If the tap is active, return the promiseVow to the caller.  This
+            // will delay the middleware until fulfilment of the retP chain.
+            return retP;
+          }
+
+          // Otherwise, passively return nothing.
+        },
+      },
+      encodeAckWatcher: {
+        onFulfilled(rawAck, { ackMethod }) {
+          // Encode the tap's ack and write it out.
+          const ack = dataToBase64(coerceToData(rawAck));
+          ackMethod = { ...ackMethod, ack };
+          return E(this.state.system).downcall(ackMethod);
+        },
+      },
+      sendErrorWatcher: {
+        onRejected(error, { ackMethod }) {
+          console.error(`Error sending ack:`, error);
+          const rawAck = JSON.stringify({ error: error.message });
+          ackMethod = { ...ackMethod, ack: dataToBase64(rawAck) };
+          return E(this.state.system).downcall(ackMethod);
+        },
+      },
+      logErrorWatcher: {
+        onRejected(error, { obj }) {
+          console.error(`Error in handling of`, obj, error);
+        },
+      },
+    },
+  );
+
+  /** @param {Parameters<typeof makeTransferInterceptorKit>} args */
+  const makeTransferInterceptor = (...args) =>
+    makeTransferInterceptorKit(...args).public;
+  return makeTransferInterceptor;
+};
+harden(prepareTransferInterceptor);
+
+const TransferMiddlewareI = M.interface('TransferMiddleware', {
+  registerTap: M.callWhen(
+    M.string(),
+    M.remotable('TransferInterceptor'),
+  ).returns(M.remotable('TargetUnregister')),
+  registerActiveTap: M.callWhen(
+    M.string(),
+    M.remotable('TransferInterceptor'),
+  ).returns(M.remotable('TargetUnregister')),
+  unregisterTap: M.callWhen(M.string()).returns(),
```

Comment:
What is a "TargetUnregister"? Shouldn't this use the debug name from packages/vats/src/bridge-target.js?